nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/authorization/AuthorizeControllerServiceReference.java

```diff
@@ -109,17 +120,54 @@ public static void authorizeControllerServiceReferences(final Map<String, String
                             final ParameterTokenList tokenList = parser.parseTokens(proposedValue);
                             final boolean referencesParameter = !tokenList.toReferenceList().isEmpty();
                             if (referencesParameter) {
-                                throw new IllegalArgumentException("The property '" + propertyDescriptor.getDisplayName() + "' cannot reference a Parameter because the property is a " +
-                                    "Controller Service reference. Allowing Controller Service references to make use of Parameters could result in security issues and a poor user experience. " +
-                                    "As a result, this is not allowed.");
+                                authorizeControllerServiceReference(authorizable, authorizer, lookup, user, propertyDescriptor, proposedEffectiveValue);
+                            } else {
+                                final Authorizable newServiceAuthorizable = lookup.getControllerService(proposedValue).getAuthorizable();
+                                newServiceAuthorizable.authorize(authorizer, RequestAction.READ, user);
                             }
-
-                            final Authorizable newServiceAuthorizable = lookup.getControllerService(proposedValue).getAuthorizable();
-                            newServiceAuthorizable.authorize(authorizer, RequestAction.READ, user);
                         }
                     }
                 }
             }
         }
     }
+
+    /**
+     * Authorizes a proposed new controller service reference.
+     *
+     * @param authorizable authorizable that may reference a controller service
+     * @param authorizer authorizer
+     * @param lookup lookup
+     * @param user user
+     * @param propertyDescriptor the propertyDescriptor referencing a controller service
+     * @param proposedEffectiveValue the new proposed value (id of the controller service) to use as a reference
+     */
+    public static void authorizeControllerServiceReference(
+        ComponentAuthorizable authorizable,
+        Authorizer authorizer,
+        AuthorizableLookup lookup,
+        NiFiUser user,
+        PropertyDescriptor propertyDescriptor,
+        String proposedEffectiveValue
+    ) {
+        final String currentValue = authorizable.getValue(propertyDescriptor);
+
+        if (authorizable.getAuthorizable() instanceof ComponentNode) {
+            authorize(authorizer, lookup, user, currentValue);
+            authorize(authorizer, lookup, user, proposedEffectiveValue);
+        } else {
+            throw new IllegalArgumentException("The property '" + propertyDescriptor.getDisplayName() + "' cannot reference a Parameter because the property is a " +
+                "Controller Service reference. Allowing Controller Service references to make use of Parameters could result in security issues and a poor user experience. " +
+                "As a result, this is not allowed.");
```

Comment:
To be honest it should be impossible to reach this clause but I guess in case it happens for some reason there should be a better error message.
I'll change it.